File: packages/snow_draw_core/lib/draw/types/snap_guides.dart
import 'package:meta/meta.dart';

import '../utils/list_equality.dart';
import 'draw_point.dart';

enum SnapGuideKind { point, gap }

enum SnapGuideAxis { horizontal, vertical }

/// Visual guide information for snapping.
@immutable
class SnapGuide {
  const SnapGuide({
    required this.kind,
    required this.axis,
    required this.start,
    required this.end,
    this.markers = const [],
    this.label,
  });
  final SnapGuideKind kind;
  final SnapGuideAxis axis;
  final DrawPoint start;
  final DrawPoint end;
  final List<DrawPoint> markers;
  final double? label;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is SnapGuide &&
          other.kind == kind &&
          other.axis == axis &&
          other.start == start &&
          other.end == end &&
          pointListEquals(other.markers, markers) &&
          other.label == label;

  @override
  int get hashCode =>
      Object.hash(kind, axis, start, end, Object.hashAll(markers), label);
}

/// Element-wise equality for [SnapGuide] lists.
bool snapGuideListEquals(List<SnapGuide> a, List<SnapGuide> b) =>
    listEquals(a, b);

File: packages/snow_draw_rust_engine/src/draw/types/snap_guides.rs
#![allow(dead_code)]

use crate::draw::types::draw_point::DrawPoint;
use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum SnapGuideKind {
    Point,
    Gap,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum SnapGuideAxis {
    Horizontal,
    Vertical,
}

/// Visual guide information for snapping.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct SnapGuide {
    pub kind: SnapGuideKind,
    pub axis: SnapGuideAxis,
    pub start: DrawPoint,
    pub end: DrawPoint,
    pub markers: Vec<DrawPoint>,
    pub label: Option<f64>,
}

impl SnapGuide {
    pub fn new(kind: SnapGuideKind, axis: SnapGuideAxis, start: DrawPoint, end: DrawPoint) -> Self {
        Self {
            kind,
            axis,
            start,
            end,
            markers: Vec::new(),
            label: None,
        }
    }
}

pub fn snap_guide_list_equals(a: &[SnapGuide], b: &[SnapGuide]) -> bool {
    a == b
}
